### ISSUE TITLE
How to save committed changes in io cache to current file

### ISSUE THREAD

**Issue by A1phaZer0** [radare/radare2#8388]
radare2 1.7.0-git 15927 @ linux-x86-32 git.1.6.0-654-gd67ecce
commit: d67ecceea17bfecd5f0b3806f33dbea57ced1eaf build: 2017-08-29__18:30:48

Hi,
It seems that r2 can't save committed changes to current file when io.cache=true, no matter with -w option or not, is r2 designed that way or I just misunderstand something. Also, there's some kind of redundant information(cached), like 46->46, 49->49 below.

```c
[0x08048440]> x/4xb
- offset -   0 1  2 3  4 5  6 7  8 9  A B  C D  E F  0123456789ABCDEF
0x08048440  4641 4142                                FAAB
[0x08048440]> woa 03 @ !1
[0x08048440]> x/4xb
- offset -   0 1  2 3  4 5  6 7  8 9  A B  C D  E F  0123456789ABCDEF
0x08048440  4941 4142                                IAAB
[0x08048440]> wci
[0x08048440]> wc
idx=0 addr=0x08048440 size=1 46 -> 49 (written)
idx=1 addr=0x08048440 size=1 49 -> 49 (not written)
[0x08048440]> woa 05 @0x8048441 !1
[0x08048440]> x/4xb
- offset -   0 1  2 3  4 5  6 7  8 9  A B  C D  E F  0123456789ABCDEF
0x08048440  4946 4142                                IFAB
[0x08048440]> wci
[0x08048440]> wc
idx=0 addr=0x08048440 size=1 46 -> 49 (written)
idx=1 addr=0x08048440 size=1 49 -> 49 (not written)
idx=2 addr=0x08048441 size=1 41 -> 46 (not written)
idx=3 addr=0x08048440 size=1 49 -> 49 (not written)
[0x08048440]> wc+ 0x8048441
[0x08048440]> wc
idx=0 addr=0x08048440 size=1 46 -> 49 (written)
idx=1 addr=0x08048440 size=1 49 -> 49 (not written)
idx=2 addr=0x08048441 size=1 41 -> 46 (written)
idx=3 addr=0x08048440 size=1 49 -> 49 (not written)
idx=4 addr=0x08048441 size=1 46 -> 46 (not written)
```

**Comment by condret** [comment]
what does the om say?

**Comment by A1phaZer0** [comment]
It's: 
```c
[0x08048440]> om
 3 fd: 3 +0x00000000 0x08048000 - 0x0804887f -r-x fmap.LOAD0
 2 fd: 3 +0x00000f08 0x08049f08 - 0x0804a02b -r-- fmap.LOAD1
 1 fd: 4 +0x00000000 0x0804a02c - 0x0804a02f mrw- mmap.LOAD1
```

**Comment by radare** [comment]
The wci command should do that

**Comment by A1phaZer0** [comment]
> The wci command should do that

Hi, I tried `wci` and after that `q` or `o--` to quit r2\close file, but nothing saved.

**Comment by condret** [comment]
try this: ````SRbp 0; wci```` we just replaced the old io layer, with a new one that is more focused on correctness, which is needed for accurate emulation. If the mapped area of the file says, it is not writeable, the new io, won't let you write. You can reapply the sections of a binary with SRbp for patching-mode instead of analysis mode. See SR?

**Comment by MaskRay** [comment]
I understand pancake is stubborn on using `bool r_io_read_at`, but I think you can at least make `st64/int r_io_write_at` and return `ENOSPC/EFBIG/EIO/EPERM` and give the user a friendly error message when a write fails, instead of silently swallowing the error.

**Comment by A1phaZer0** [comment]
@condret 
I use `r2 -e io.cache=true -w` to open the file again trying `wci` or `SRbp 0; wci` but still the same story. 
```c
[0x08048440]> om
 3 fd: 3 +0x00000000 0x08048000 - 0x0804887f -rwx fmap.LOAD0
 2 fd: 3 +0x00000f08 0x08049f08 - 0x0804a02b -rw- fmap.LOAD1
 1 fd: 4 +0x00000000 0x0804a02c - 0x0804a02f mrw- mmap.LOAD1
[0x08048440]> SRbp 0; wci
[0x08048440]> wc
idx=0 addr=0x08048440 size=1 46 -> 49 (written)
idx=1 addr=0x08048440 size=1 49 -> 49 (not written)
[0x08048440]> q
```

**Comment by MaskRay** [comment]
Bugs aside, I think there are other off-by-1 here
https://github.com/radare/radare2/pull/8402 the sooner thebetter.

**Comment by condret** [comment]
this will stay bool, as long as r2con ppl decide different

**Comment by basicinfosecurity** [comment]
Hi, has this been fixed? I've been using 3.2.1 and cannot seem to save modifications using wci after setting `io.cache`, `io.cache.auto`, `io.cache.read` & `io.cache.write` to true. I've also tried opening binary in write mode but I'm also not having any luck. Any tips? I'm trying to add a new section using `om` and saving using `wci`.
